src/wasm-lib/kcl/src/parser/parser_impl.rs

```diff
@@ -966,7 +967,16 @@ fn body_items_within_function(i: TokenSlice) -> PResult<WithinFunction> {
             non_code_node.map(WithinFunction::NonCode)
         },
         _ =>
-            (expression_stmt.map(BodyItem::ExpressionStatement), opt(noncode_just_after_code)).map(WithinFunction::BodyItem),
+            alt((
+                (
+                    declaration.map(BodyItem::VariableDeclaration),
+                    opt(noncode_just_after_code)
+                ).map(WithinFunction::BodyItem),
+                (
+                    expression_stmt.map(BodyItem::ExpressionStatement),
+                    opt(noncode_just_after_code)
+                ).map(WithinFunction::BodyItem),
+            ))
```

Comment:
It looks like this code tries to parse a variable declaration, and if it fails, it backtracks and tries to parse an expression statement.

I think this works because we don't have assignment to arbitrary things (like `arr[i][j + 1] = foo`). If we did, this code would backtrack quite a bit, which would be bad. For us, a variable declaration can only ever look forward two tokens, an identifier and an equals sign.

So I think this is okay. The use of `alt()` with anything other than terminals is really easy to accidentally introduce performance problems, so I'm always wary.

(Aside: The way Crafting Interpreters does it is by always parsing an expression and then looking for a trailing `=`. If it finds it, it knows that the expression it just parsed is the place [AKA l-value] of an assignment.)

Forgive me if you've already thought through all of this 😅

Reply:
Thanks for pointing it out! I thought about the perf cost for the current parser, but I hadn't though about assignment to arbitrary expressions like that. But, we don't have mutation/rebinding so it should be OK :)